---
title: "QC and clustering on ISCs"
author: "Jordi Camps"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---

```{r message=FALSE}
library(SingleCellExperiment)
library(scater)
library(scTools)
library(edgeR)
library(ggthemes)
library(ggExtra)
library(SC3)
library(ggplot2)
library(dplyr)
library(scran)
library(viridis)
library(RColorBrewer)
set.seed(1234567)
```

## Import data set
```{r}
sce <- readRDS("data/reads_qc_scran_sc3.rds")
#filter out ISCs (cluster 1:3 for k = 7)
sce <- sce[, c(sce$sc3_7_clusters == "1" | sce$sc3_7_clusters == "2" | sce$sc3_7_clusters == "3")]
#reducedDim(sce) <- NULL
sce
```

```{r fig.height=4, fig.width=5}
plotPCA(sce, colour_by = "sc3_7_clusters")
plotPCA(sce, colour_by = "Cd55")
plotPCA(sce, colour_by = "Gdf10")
plotPCA(sce, colour_by = "Thbs4")
```


Change factors
```{r}
sce$sc3_7_clusters <- plyr::revalue(sce$sc3_7_clusters, c("1" = "ISC 3", "2" = "ISC 1", "3" = "ISC 2"))
sce$Cluster <- factor(sce$sc3_7_clusters, levels = c("ISC 1", "ISC 2", "ISC 3"))
sce$genotype <- factor(sce$genotype, levels = c("Healthy", "Dystrophic"))
```

plot PCA
```{r fig.height=3, fig.width=4}
reducedDim(sce) <- NULL
sce <- plotPCA(sce, colour_by = "Cluster", return_SCE = TRUE, ncomponents = 2)
sce$PC1 <- reducedDim(sce)[, 1]
sce$PC2 <- reducedDim(sce)[, 2]
plot_dims(sce_object = sce, color = "Cluster")
```

Plot detected genes
```{r fig.height=3, fig.width=2.25}
ggplot(as_tibble(colData(sce)), aes(y = total_features, x = Cluster, fill = Cluster)) +
  geom_boxplot(outlier.shape = TRUE, notch = FALSE, varwidth = FALSE) +
  #geom_violin() +
  #ggbeeswarm::geom_quasirandom(width = 0.4, size = 2) +
  scale_fill_brewer(type = "qual", palette = "Set2") +
  #scale_color_brewer(type = "qual", palette = "Set2") +
  scale_y_continuous("Number of genes per cell") +
  theme_bw(base_size = 14) +
  theme(panel.border = element_blank(), strip.background = element_blank(), legend.position = "none", 
        axis.line = element_line(colour = "black", size = 0.5), strip.text.x = element_text(size = 24), 
        axis.ticks.y = element_line(colour = "black", size = 0.5), axis.text = element_text(colour = "black"), panel.grid = element_blank(),
        axis.title.x = element_blank(), axis.text.x = element_text(angle = 45, hjust = 1))
ggsave("plots/ISC/detected_genes.pdf", dpi = 600, width = 2.25, height = 3)
ggsave("plots/ISC/detected_genes.tiff", dpi = 600, width = 2.25, height = 3)
```

## QC
Total counts 
```{r fig.height=3, fig.width=3}
ggplot(as.data.frame(colData(sce)), aes(total_counts / 100000)) +
  geom_histogram(bins = 20) +
  geom_vline(xintercept = 1.5, color = "red", size = 1.5) +
  scale_x_continuous(paste("Aligned reads/cell (", 10^5, ")", sep = ""), expand = c(0, 0)) +
  scale_y_continuous("Number of cells", expand = c(0, 0)) +
  theme_bw(base_size = 18) +
  theme(panel.border = element_blank(), panel.grid = element_blank(), axis.line = element_line(colour = "black"), 
      axis.ticks = element_line(colour = "black"), axis.text = element_text(colour = "black"), axis.title.x = element_text(hjust = 0.9))
ggsave("plots/QC/ISC/reads_per_cell.pdf", width = 3, height = 3, dpi = 600)
```

Total genes detected
```{r fig.height=3, fig.width=3}
ggplot(as.data.frame(colData(sce)), aes(total_features / 1000)) +
  geom_histogram(bins = 20) +
  geom_vline(xintercept = c(1, 6), color = "red", size = 1.5) +
  scale_x_continuous(paste("Detected genes (", 10^3, ")", sep = ""), expand = c(0, 0)) +
  scale_y_continuous("Number of cells", expand = c(0, 0)) +
  theme_bw(base_size = 18) +
  theme(panel.border = element_blank(), panel.grid = element_blank(), axis.line = element_line(colour = "black"), 
      axis.ticks = element_line(colour = "black"), axis.text = element_text(colour = "black"), axis.title.x = element_text(hjust = 0.85))
ggsave("plots/QC/ISC/genes_per_cell.pdf", width = 3, height = 3, dpi = 600)
```


Gene filtering
```{r}
filter_genes <- apply(counts(sce), 1, 
                      function(x) length(x[x > 1]) >= 2)
rowData(sce)$use <- filter_genes

knitr::kable(
  as.data.frame(table(filter_genes)),
  booktabs = TRUE,
  row.names = FALSE,
  caption = 'The number of genes removed by gene filter (FALSE)'
)
```

mtDNA
```{r fig.height=3, fig.width=3}
ggplot(as.data.frame(colData(sce)), aes(pct_counts_MT)) +
  geom_histogram(bins = 20) +
  geom_vline(xintercept = 6, color = "red", size = 1.5) +
  #facet_wrap(~batch) +
  scale_x_continuous("% mitochrondrial genes", expand = c(0, 0)) +
  scale_y_continuous("Number of cells", expand = c(0, 0)) +
  theme_bw(base_size = 18) +
  theme(panel.border = element_blank(), panel.grid = element_blank(), axis.line = element_line(colour = "black"), 
      axis.ticks = element_line(colour = "black"), axis.text = element_text(colour = "black"), axis.title.x = element_text(hjust = 0.85))
ggsave("plots/QC/ISC/mtDNA_per_cell.pdf", width = 3, height = 3, dpi = 600)

filter_by_MT <- sce$pct_counts_MT < 6
knitr::kable(
  as.data.frame(table(filter_by_MT)),
  booktabs = TRUE,
  row.names = FALSE,
  caption = 'The number of cells removed by MT filter (FALSE)'
)
```

ERCC
```{r fig.height=3, fig.width=3}
ggplot(as.data.frame(colData(sce)), aes(pct_counts_ERCC)) +
  geom_histogram(bins = 20) +
  geom_vline(xintercept = 5.5, color = "red", size = 1.5) +
  #facet_wrap(~batch) +
  scale_x_continuous("% ERCCs", expand = c(0, 0)) +
  scale_y_continuous("Number of cells", expand = c(0, 0)) +
  theme_bw(base_size = 18) +
  theme(panel.border = element_blank(), panel.grid = element_blank(), axis.line = element_line(colour = "black"), 
      axis.ticks = element_line(colour = "black"), axis.text = element_text(colour = "black"), axis.text.x = element_text(hjust = 0.6))
ggsave("plots/QC/ISC/ERCCs_per_cell.pdf", width = 3, height = 3, dpi = 600)

filter_by_ERCC <- (sce$pct_counts_ERCC < 5.5)

knitr::kable(
  as.data.frame(table(filter_by_ERCC)),
  booktabs = TRUE,
  row.names = FALSE,
  caption = 'The number of cells removed by ERCC filter (FALSE)'
)
```

Count good and bad libraries
```{r}
sce$use <- (
    # sufficient endogenous RNA
    filter_by_ERCC &
    # remove cells with unusual number of reads in MT genes
    filter_by_MT
)
table(sce$use)
```

Filter out bad libraries
```{r}
sce <- sce[rowData(sce)$use, colData(sce)$use]
```


# Technical components
For total features Component 1 does only have an R-squared of 0.1. Technical components is PC7.
```{r fig.height=4, fig.width=4}
plotQC(sce, type = "find-pcs", variable = "total_counts")
plotQC(sce, type = "find-pcs", variable = "total_features")
```

Plot PCA on previous clusters  and genotype
```{r fig.height=4, fig.width=5}
reducedDim(sce) <- NULL
plotPCA(sce, colour_by = "Cluster", ncomponents = 4)
plotPCA(sce, colour_by = "genotype", ncomponents = 4)
```

Export PCAs
```{r fig.height=4, fig.width=5}
sce = plotPCA(sce, ncomponents = 4, colour_by = "Cluster", return_SCE = TRUE)
sce$PC1 <- reducedDim(sce)[,1]
sce$PC2 <- reducedDim(sce)[,2]
sce$PC3 <- reducedDim(sce)[,3]
sce$PC3 <- reducedDim(sce)[,4]
```

Plot PCA on three clusters
```{r fig.height=3.5, fig.width=3}
plot_dims(sce_object = sce, x = "PC1", y = "PC2", color = "Cluster", point_size = 3, theme = 16) +
  scale_color_brewer(type = "qual", palette = "Set2") +
  labs(x = "PC1", y = "PC2") +
  guides(shape = FALSE) +
  theme(legend.position = "top", legend.title = element_blank())
#ggMarginal(p, groupColour = TRUE, groupFill = TRUE)
ggsave("plots/ISC/ISC_PCA.pdf", dpi = 600, width = 3, height = 3.5)
ggsave("plots/ISC/ISC_PCA.tiff", dpi = 600, width = 3, height = 3.5)
```

# Clustering
```{r}
sce <- sc3_prepare(sce)
sce <- sc3_estimate_k(sce)
metadata(sce)$sc3$k_estimation
sce <- sc3(sce, ks = 3:5, biology = TRUE)
```

```{r eval=FALSE, include=FALSE}
sc3_interactive(sce)
```

Export results to excel
**Install perl on your computer to use this function!**
```{r}
#sc3_export_results_xls(sce)
library(xlsx)
write.xlsx(rowData(sce), file = "tables/ISC/genes_sc3_results.xlsx", col.names = TRUE)
write.xlsx(colData(sce), file = "tables/ISC/cells_sc3_results.xlsx", col.names = TRUE)
```

### Export csv with all differentially expressed genes
k = 5
```{r}
#sc3_plot_de_genes(sce, k = 5, show_pdata = TRUE)

table <- sce %>%
  rowData() %>%
  as_tibble() %>%
  filter(sc3_5_de_padj < 0.05) %>%
  select(feature_symbol, ensembl_gene_id, sc3_5_markers_auroc, sc3_5_markers_clusts, sc3_5_markers_padj, sc3_5_de_padj) %>%
  arrange(sc3_5_markers_clusts, sc3_5_markers_auroc)

write.xlsx(table, file = "tables/ISC/k5_de_genes.xlsx")
```

Save clustering results
```{r}
saveRDS(sce, "data/reads_qc_scran_ISC_SC3.rds")
```

