---
title: "Clustering with SC3"
author: "Jordi Camps"
date: "`r format(Sys.time(), '%d %B, %Y, %H:%M')`"
output: html_document
---

# Load libraries
```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
library(SC3)
library(scater)
library(RColorBrewer)
library(viridis)
library(SingleCellExperiment)
library(grid)
library(BiocGenerics)
library(scTools)
library(dplyr)
library(plyr)
library(lazyeval)
library(tidyr)
library(scran)
library(ggthemes)
library(xlsx)
options(stringsAsFactors = FALSE)
```

# Import dataset
```{r}
set.seed(123456)
sce <- readRDS("data/reads_qc_ERCC_strict.rds")
endog_genes <- !rowData(sce)$is_feature_control
sce
```


Normalize
```{r fig.height=3, fig.width=3}
qclust <- quickCluster(sce, min.size = 30)
sce <- computeSumFactors(sce, cluster = qclust)
sce <- computeSpikeFactors(sce, general.use = FALSE)
summary(sizeFactors(sce))
plot(sizeFactors(sce), sce$total_counts/1e3, log = "xy", ylab = "Library size (thousands)", xlab = "Size factor")
sce <- normalize(sce)

saveRDS(sce, "data/reads_qc_ERCC_strict_scran.rds")
```

```{r fig.height=3, fig.width=3}
sce$size_factor <- sizeFactors(sce)
ggplot(as.data.frame(colData(sce)), aes(y = size_factor, x = total_counts / 100000)) +
  geom_point() +
  geom_smooth(method = "lm", color = "red") +
  scale_y_continuous("Scran size factor", expand = c(0, 0)) +
  scale_x_continuous(paste("Total reads / cell (", 10^5, ")", sep = ""), expand = c(0, 0)) +
  theme_bw(base_size = 18) +
  theme(panel.border = element_blank(), panel.grid = element_blank(), axis.line = element_line(colour = "black"),
        axis.ticks = element_line(colour = "black"), axis.text = element_text(colour = "black"), axis.title.x = element_text(hjust = 0.9))
ggsave("plots/QC/scran_size_factor.pdf", dpi = 600, width = 3, height = 3)
```


```{r}
sce$genotype <- factor(sce$genotype, levels = c("Healthy", "Dystrophic"))
sce <- sce[endog_genes, ]
sce
```

# plot QC
```{r}
plotQC(sce, type = "find-pcs", variable = "total_counts")
plotQC(sce, type = "find-pcs", variable = "total_features")
```

## Explanatory variables
```{r fig.height=3, fig.width=4}
plotQC(
    sce, type = "expl", exprs_values = "logcounts", variables = c(
      "total_features",
      "total_counts",
      "batch",
      "genotype",
      "marker",
      "pct_counts_ERCC",
      "pct_counts_MT"))
```

## Marker genes
```{r fig.height=2.5, fig.width=3}
df <- as.data.frame(logcounts(sce))
df <- tibble::rownames_to_column(df, var = "gene")
df <- tidyr::gather(df, key = "cell", value = "logcounts", -gene)
df <- dplyr::filter(df, gene == "Actb" | gene == "Ptprc" | gene == "Pecam1" | gene == "Kdr" | gene == "Adgre1" | gene == "Kit" | gene == "Flt3" | gene == "Cdh5")
ptprc <- data.frame(gene = "Ptprc", cell = unique(df$cell), logcounts = rep(0, 256))
flt3 <- data.frame(gene = "Flt3", cell = unique(df$cell), logcounts = rep(0, 256))
df <- rbind(df, ptprc, flt3)
df$gene <- plyr::revalue(df$gene, replace = c(Actb = "Actb", Ptprc = "Ptprc (Cd45)", Pecam1 = "Pecam1 (Cd31)", Kdr = "Kdr (Vegfr2)",
                                              Kit = "Kit (C-kit)", Flt3 = "Flt3 (Flk2)"))
df$gene <- factor(df$gene, levels = c("Actb", "Ptprc (Cd45)", "Kit (C-kit)", "Flt3 (Flk2)", "Pecam1 (Cd31)", "Kdr (Vegfr2)", "Cdh5"))
ggplot(df, aes(x = gene, y = logcounts, fill = gene)) +
  geom_boxplot() +
  scale_y_continuous("Normalized logcounts", expand = c(0, 0)) +
  theme_bw(base_size = 16) +
  theme(panel.border = element_blank(), panel.grid = element_blank(), axis.line = element_line(colour = "black"), 
        axis.ticks.x = element_blank(), axis.ticks.y = element_line(colour = "black"), axis.text.y = element_text(colour = "black"),
        axis.title.x = element_blank(), legend.position = "none", axis.text.x = element_text(size = 12, colour = "black", angle = 45, hjust = 1),
        axis.title.y = element_text(hjust = 1))
ggsave("plots/QC/endothelial_hematopoietic_genes.pdf", dpi = 600, width = 3, height = 2.5)
```



# plot PCA
```{r fig.height=6, fig.width=6}
plotPCA(sce, ncomponents = 4, colour_by = "marker")
plotPCA(sce, ncomponents = 4, colour_by = "genotype")
```

# Clustering
## SC3
```{r, message = FALSE, warning=FALSE}
sce <- sc3_prepare(sce)
sce <- sc3_estimate_k(sce)
metadata(sce)$sc3$k_estimation
sce <- sc3(sce, ks = 4:9, biology = TRUE, rand_seed = 123456)
```

### SC3 interactive
Check dataset interactively online
```{r eval=FALSE, include=FALSE}
sc3_interactive(sce)
```

### Create column clusters with cluster names
```{r}
sce$Clusters <- plyr::revalue(sce$sc3_7_clusters, c("1" = "ISC 3", "2" = "ISC 1", "3" = "ISC 2", "4" = "Satellite cells", "5" = "Schwann cells", 
                                                    "6" = "Smooth muscle cells", "7" = "Fibroblasts"))
```

### Export results to excel
**Install perl on your computer to use this function!**
```{r}
sc3_export_results_xls(sce)
```

```{r}
saveRDS(sce, "data/reads_qc_scran_sc3.rds")
```