---
title: "Quality control"
author: "Jordi Camps"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---

```{r setup, include=FALSE}
library(scater)
library(knitr)
library(SingleCellExperiment)
library(dplyr)
library(ggplot2)
library(scran)
options(stringsAsFactors = FALSE)
```

# Prepare dataset
## Load counts and annotation file
```{r}
reads <- read.csv("data/htseq_counts_all_mtDNA.csv")
rownames(reads) <- reads[, 1]
reads <- reads[, -1]
anno <- read.csv("data/annotation.csv", header = TRUE)
i5 <- read.csv("data/i5_indexes.csv", header = TRUE)
i7 <- read.csv("data/i7_indexes.csv", header = TRUE)
anno <- anno %>% 
  merge(i7, by = "i7_index") %>%
  merge(i5, by = "i5_index")
anno$sample_id <- paste(anno$i7_base, ".", anno$i5_base)
anno <- anno[order(anno$sample_id),] #order according to sample_id
```

### Inspect expression matrix
```{r}
head(reads[ , 1:3])
```

### Inspect annotation matrix
```{r}
head(anno)
```

### Create SingleCellExperiment object
```{r}
reads <- SingleCellExperiment(assays = list(counts = as.matrix(reads)), colData = anno)
```

### Remove genes that are not expressed in any cell
```{r}
keep_feature <- rowSums(counts(reads) > 0) > 0
reads <- reads[keep_feature, ]
```

### Define mtDNA and ERCC
```{r}
isSpike(reads, "ERCC") <- grepl("^ERCC-", rownames(reads))
isSpike(reads, "MT") <- rownames(reads) %in% 
  c("ENSMUSG00000064336", "ENSMUSG00000064337", "ENSMUSG00000064338",
    "ENSMUSG00000064339", "ENSMUSG00000064340", "ENSMUSG00000064341",
    "ENSMUSG00000064342", "ENSMUSG00000064343", "ENSMUSG00000064344",
    "ENSMUSG00000064345", "ENSMUSG00000064346", "ENSMUSG00000064347", 
    "ENSMUSG00000064348", "ENSMUSG00000064349", "ENSMUSG00000064350",
    "ENSMUSG00000064351", "ENSMUSG00000064352", "ENSMUSG00000064353",
    "ENSMUSG00000064354", "ENSMUSG00000064355", "ENSMUSG00000064356",
    "ENSMUSG00000064357", "ENSMUSG00000064358", "ENSMUSG00000064359",
    "ENSMUSG00000064360", "ENSMUSG00000064361", "ENSMUSG00000065947",
    "ENSMUSG00000064363", "ENSMUSG00000064364", "ENSMUSG00000064365",
    "ENSMUSG00000064366", "ENSMUSG00000064367", "ENSMUSG00000064368",
    "ENSMUSG00000064369", "ENSMUSG00000064370", "ENSMUSG00000064371",
    "ENSMUSG00000064372")

reads <- calculateQCMetrics(reads,
  feature_controls = list(ERCC = isSpike(reads, "ERCC"), MT = isSpike(reads, "MT"))
)
```


# Cell quality control
## Total counts per cell
```{r fig.height=3, fig.width=3}
ggplot(as.data.frame(colData(reads)), aes(total_counts / 100000)) +
  geom_histogram(bins = 30) +
  geom_vline(xintercept = 1.5, color = "red", size = 1.5) +
  scale_x_continuous(paste("Aligned reads/cell (", 10^5, ")", sep = ""), expand = c(0, 0)) +
  scale_y_continuous("Number of cells", expand = c(0, 0)) +
  theme_bw(base_size = 18) +
  theme(panel.border = element_blank(), panel.grid = element_blank(), axis.line = element_line(colour = "black"), 
      axis.ticks = element_line(colour = "black"), axis.text = element_text(colour = "black"), axis.title.x = element_text(hjust = 0.9))
ggsave("plots/QC/reads_per_cell.pdf", width = 3, height = 3, dpi = 600)
```


```{r}
filter_by_total_counts <- reads$total_counts > 1.5e5

table(filter_by_total_counts)
```


## Detected genes per cell
```{r fig.height=3, fig.width=3}
ggplot(as.data.frame(colData(reads)), aes(total_features / 1000)) +
  geom_histogram(bins = 30) +
  geom_vline(xintercept = c(1, 6), color = "red", size = 1.5) +
  scale_x_continuous(paste("Detected genes (", 10^3, ")", sep = ""), expand = c(0, 0)) +
  scale_y_continuous("Number of cells", expand = c(0, 0)) +
  theme_bw(base_size = 18) +
  theme(panel.border = element_blank(), panel.grid = element_blank(), axis.line = element_line(colour = "black"), 
      axis.ticks = element_line(colour = "black"), axis.text = element_text(colour = "black"), axis.title.x = element_text(hjust = 0.85))
ggsave("plots/QC/genes_per_cell.pdf", width = 3, height = 3, dpi = 600)
```


```{r}
filter_by_expr_features <- (reads$total_features > 1000 & reads$total_features < 6000)

table(filter_by_expr_features)
```

## mtDNA
```{r, fig.height=4.5, fig.width=6}
plotPhenoData(reads, x = "total_features", y = "pct_counts_MT", colour_by = "batch") +
  geom_hline(yintercept = 6, col = "red") +
  labs(x = "Total genes", y = "Percentage mitochondrial genes")
```

percentage mtDNA per cell
```{r fig.height=3, fig.width=3}
ggplot(as.data.frame(colData(reads)), aes(pct_counts_MT)) +
  geom_histogram(bins = 50) +
  geom_vline(xintercept = 6, color = "red", size = 1.5) +
  #facet_wrap(~batch) +
  scale_x_continuous("% mitochrondrial genes", expand = c(0, 0)) +
  scale_y_continuous("Number of cells", expand = c(0, 0)) +
  theme_bw(base_size = 18) +
  theme(panel.border = element_blank(), panel.grid = element_blank(), axis.line = element_line(colour = "black"), 
      axis.ticks = element_line(colour = "black"), axis.text = element_text(colour = "black"), axis.title.x = element_text(hjust = 0.85))
ggsave("plots/QC/mtDNA_per_cell.pdf", width = 3, height = 3, dpi = 600)
```


```{r}
filter_by_MT <- reads$pct_counts_MT < 6

table(filter_by_MT)
```


## ERCC
```{r}
plotPhenoData(reads, x = "total_features", y = "pct_counts_ERCC", colour_by = "batch") +
  geom_hline(yintercept = 10, col = "red") +
  labs(x = "Total genes", y = "Percentage ERCC") 

plotPhenoData(reads, x = "total_features", y = "pct_counts_ERCC", colour_by = "marker") +
  geom_hline(yintercept = 10, col = "red") +
  labs(x = "Total genes", y = "Percentage ERCC") 
```

percentage ERCC per cell
```{r fig.height=3, fig.width=3}
ggplot(as.data.frame(colData(reads)), aes(pct_counts_ERCC)) +
  geom_histogram(bins = 60) +
  geom_vline(xintercept = 10, color = "red", size = 1.5) +
  #facet_wrap(~batch) +
  scale_x_continuous("% ERCCs", expand = c(0, 0)) +
  scale_y_continuous("Number of cells", expand = c(0, 0)) +
  theme_bw(base_size = 18) +
  theme(panel.border = element_blank(), panel.grid = element_blank(), axis.line = element_line(colour = "black"), 
      axis.ticks = element_line(colour = "black"), axis.text = element_text(colour = "black"), axis.text.x = element_text(hjust = 0.6))
ggsave("plots/QC/ERCCs_per_cell.pdf", width = 3, height = 3, dpi = 600)
```


```{r}
filter_by_ERCC <- (reads$pct_counts_ERCC < 10)

table(filter_by_ERCC)
```

## Cell cycle
```{r}
mm.pairs <- readRDS(system.file("exdata", "mouse_cycle_markers.rds", package="scran"))
assigned <- cyclone(reads, pairs=mm.pairs)
head(assigned$scores)
table(assigned$phases)
table(assigned$phases, reads$genotype)


filter_by_cell_cycle <- (assigned$phases == "G1")
```

```{r fig.height=3, fig.width=3}
df <- data_frame(Phase = assigned$phases, genotype = reads$genotype)
df$Phase <- factor(df$Phase, levels = c("G2M", "S", "G1"))
df$genotype <- factor(df$genotype, levels = c("Healthy", "Dystrophic"))
ggplot(df, aes(x = genotype, fill = Phase)) +
  geom_bar(width = 0.8) +
  scale_fill_manual(values = rev(RColorBrewer::brewer.pal(3, "Set1"))) +
  scale_y_continuous("Number of cells" , expand = c(0, 0)) +
  scale_x_discrete(labels = c("Healthy" = "H", "Dystrophic" = "D")) +
  theme_bw(base_size = 18) +
  theme(panel.border = element_blank(), panel.grid = element_blank(), axis.ticks.x = element_blank(), axis.ticks.y = element_line(colour = "black"),
        axis.line = element_line(colour = "black"), axis.text = element_text(colour = "black"), 
        axis.text.y = element_text(angle = 90, hjust = 0.5), legend.position = "right", legend.title = element_blank())
ggsave("plots/QC/cell_cycle_phase_per_cell.pdf", dpi = 600, width = 3, height = 3)
```



# Cell filtering
## Manual filter
```{r}
reads$use <- (
  # sufficient features (genes)
  filter_by_expr_features &
    # sufficient molecules counted
    filter_by_total_counts &
    # sufficient endogenous RNA
    filter_by_ERCC &
    # remove cells with unusual number of reads in MT genes
    filter_by_MT &
    # remove cells that are not in G1 phase
    filter_by_cell_cycle
)
table(reads$use)
```

## Automatic filter
```{r eval=FALSE, include=FALSE}
reads <- plotPCA(
    reads,
    size_by = "total_features", 
    shape_by = "use",
    pca_data_input = "pdata",
    detect_outliers = TRUE,
    return_SCE = TRUE
)
table(reads$outlier)
```

# Gene analysis
## Annotate genes
```{r}
reads <- getBMFeatureAnnos(
  reads,
  filters = "ensembl_gene_id",
  attributes = c(
    "ensembl_gene_id",
    "mgi_symbol",
    "chromosome_name",
    "transcript_biotype",
    "transcript_start",
    "transcript_end",
    "transcript_count"),
  feature_symbol = "mgi_symbol",
  feature_id = "ensembl_gene_id",
  biomart = "ENSEMBL_MART_ENSEMBL",
  dataset = "mmusculus_gene_ensembl",
  host = "www.ensembl.org"
)
rownames(reads) <- make.names(rowData(reads)$mgi_symbol, unique = TRUE)
rownames(reads) <- gsub("NA", "ERCC", rownames(reads))
```

## Gene expression
```{r, fig.height=8, fig.width=5}
plotQC(reads, type = "highest-expression")
```

## Gene filtering
### Light gene filter (better for detecting outliers)
```{r}
filter_genes <- apply(counts(reads[, colData(reads)$use]), 1, 
                      function(x) length(x[x > 1]) >= 2)
rowData(reads)$use <- filter_genes

table(filter_genes)
```

### Harsch gene filter
```{r eval=FALSE, include=FALSE}
filter_genes <- apply(counts(reads[, colData(reads)$use]), 1, 
                      function(x) length(x[x > 1]) >= 10)
rowData(reads)$use <- filter_genes

table(filter_genes)
```


# Save the data
## Genes and cells after QC
```{r}
dim(reads[rowData(reads)$use, colData(reads)$use])
```

## Calculate raw logcounts
```{r, fig.width=5.5, fig.height=4}
assay(reads, "logcounts_raw") <- log2(counts(reads) + 1)
reducedDim(reads) <- NULL
```

## Remove bad quality libraries
```{r}
reads_qc <- reads[rowData(reads)$use, colData(reads)$use]
```

## Cells per marker and genotype left
```{r}
table(reads$marker, reads$genotype)
table(reads_qc$marker, reads_qc$genotype)
```

## Plate layout before and after QC
```{r}
temp <- as_tibble(colData(reads))
ggplot(temp, aes(x = col_XT, y = row_XT, col = marker, shape = genotype)) +
  geom_point(size = 5) +
  scale_y_discrete(limits = rev(unique(sort(temp$row_XT)))) +
  ggtitle("Plate layout before QC")
temp <- as_tibble(colData(reads_qc))
ggplot(temp, aes(x = col_XT, y = row_XT, col = marker, shape = genotype)) +
  geom_point(size = 5) +
  scale_y_discrete(limits = rev(unique(sort(temp$row_XT)))) +
  ggtitle("Plate layout after QC")
```


## Save to RDS file
```{r}
saveRDS(reads, file ="data/reads_ERCC_strict.rds")
saveRDS(reads_qc, file = "data/reads_qc_ERCC_strict.rds")
```

# Visualization
PCA and tSNE of cells before and after QC.
```{r}
reads_qc <- reads[rowData(reads)$use, colData(reads)$use]
endog_genes <- !rowData(reads_qc)$is_feature_control
```

## PCA
```{r, fig.width=11, fig.height=4}
pca1 <- plotPCA(reads[endog_genes, ],
                colour_by = "batch",
                size_by = "total_features",
                shape_by = "individual",
                exprs_values = "logcounts_raw",
                theme=12) 
pca2 <- plotPCA(reads_qc[endog_genes, ],
                colour_by = "batch",
                size_by = "total_features",
                shape_by = "individual",
                exprs_values = "logcounts_raw",
                theme=12) 
multiplot(pca1, pca2, cols = 2)
```


## tSNE
```{r, fig.width=14, fig.height=4}
tsne1 <- plotTSNE(reads[endog_genes, ],
                 perplexity = 2,
                 colour_by = "batch",
                 size_by = "total_features",
                 shape_by = "individual",
                 exprs_values = "logcounts_raw",
                 rand_seed = 123456) 
tsne2 <- plotTSNE(reads_qc[endog_genes, ],
                 perplexity = 2,
                 colour_by = "batch",
                 size_by = "total_features",
                 shape_by = "individual",
                 exprs_values = "logcounts_raw",
                 rand_seed = 123456) 
tsne3 <- plotTSNE(reads_qc[endog_genes, ],
                 perplexity = 20,
                 colour_by = "batch",
                 size_by = "total_features",
                 shape_by = "individual",
                 exprs_values = "logcounts_raw",
                 rand_seed = 123456) 
multiplot(tsne1, tsne2, tsne3, cols = 3)
```