---
title: "Clustering with SC3"
author: "Jordi Camps"
date: "`r format(Sys.time(), '%d %B, %Y, %H:%M')`"
output: html_document
---

# Load libraries
```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
library(SC3)
library(scater)
library(RColorBrewer)
library(viridis)
library(SingleCellExperiment)
library(grid)
library(BiocGenerics)
library(scTools)
library(dplyr)
library(lazyeval)
library(tidyr)
library(scran)
library(pheatmap)
library(plyr)
options(stringsAsFactors = FALSE)
```

## Import dataset
```{r}
set.seed(1234567)
sce <- readRDS("data/reads_qc.rds")
endog_genes <- !rowData(sce)$is_feature_control
sce <- sce[endog_genes, ]
```

## Normalize for library size
```{r}
qclust <- quickCluster(sce, min.size = 30)
sce <- computeSumFactors(sce, clusters = qclust)
sce <- computeSpikeFactors(sce, general.use = FALSE)
summary(sizeFactors(sce))
sce <- normalize(sce)
```

# Clustering
## SC3
```{r, message = FALSE, warning=FALSE}
sce <- sc3_prepare(sce)
sce <- sc3_estimate_k(sce)
metadata(sce)$sc3$k_estimation
sce <- sc3(sce, ks = 2:11, biology = TRUE, n_cores = 4)
```

## SC3 interactive
Check dataset interactively online
```{r eval=FALSE, include=FALSE}
sc3_interactive(sce)
```

### Save to rds
```{r}
saveRDS(sce, "data/reads_qc_scran_sc3.rds")
```

## Export results
### Export results to excel
**Install perl on your computer to use this function!**
```{r}
sc3_export_results_xls(sce)
```

### Export csv with all differentially expressed genes
k = 4
```{r}
sc3_plot_de_genes(sce, k = 4, show_pdata = TRUE)

sce %>%
  rowData() %>%
  as_tibble() %>%
  filter(sc3_4_de_padj < 0.01) %>%
  select(feature_symbol, ensembl_gene_id, sc3_4_markers_auroc, sc3_4_markers_clusts, sc3_4_de_padj) %>%
  arrange(sc3_4_de_padj) %>%
  write.csv("tables/k4_de_genes.csv")
```

## Plot results
## Export principal components and tSNE dimensions in reduced dim
```{r fig.height=10, fig.width=10}
sce = plotPCA(sce, ncomponents = 4, colour_by = "sc3_9_clusters", return_SCE = TRUE)
sce$PC1 <- reducedDim(sce)[,1]
sce$PC2 <- reducedDim(sce)[,2]
sce$PC3 <- reducedDim(sce)[,3]
sce$PC4 <- reducedDim(sce)[,4]

sce <- plotTSNE(sce, colour_by = "sc3_9_clusters", return_SCE = TRUE, perplexity = 20, exprs_values = "logcounts")
sce$Dim1 <- reducedDims(sce)$TSNE[,1]
sce$Dim2 <- reducedDims(sce)$TSNE[,2]
```

## Plot clusters on PCA and tSNE
```{r}
plot_dims(sce, "PC2", "PC3", "sc3_4_clusters")
plot_dims(sce, "Dim1", "Dim2", "sc3_4_clusters")
```

## Plot expression of clusters
```{r}
genes <- c("Dcn", "Cd82", "Plp1", "Myl9")
plot_expression(sce, genes, "sc3_4_clusters")
```

## Plot heatmap
SC3 functions
```{r}
make_col_ann_for_heatmaps <- function(object, show_pdata) {
    if (any(!show_pdata %in% colnames(colData(object)))) {
        show_pdata_excl <- show_pdata[!show_pdata %in% colnames(colData(object))]
        show_pdata <- show_pdata[show_pdata %in% colnames(colData(object))]
        message(paste0("Provided columns '", paste(show_pdata_excl, collapse = "', '"), "' do not exist in the phenoData table!"))
        if (length(show_pdata) == 0) {
            return(NULL)
        }
    }
    ann <- NULL
    if (is.null(metadata(object)$sc3$svm_train_inds)) {
        ann <- colData(object)[, colnames(colData(object)) %in% show_pdata]
    } else {
        ann <- colData(object)[metadata(object)$sc3$svm_train_inds, colnames(colData(object)) %in% 
            show_pdata]
    }
    # remove columns with 1 value only
    if (length(show_pdata) > 1) {
        keep <- unlist(lapply(ann, function(x) {
            length(unique(x))
        })) > 1
        if (!all(keep)) {
            message(paste0("Columns '", paste(names(keep)[!keep], collapse = "', '"), "' were excluded from annotation since they contained only a single value."))
        }
        ann <- ann[, names(keep)[keep]]
        if (ncol(ann) == 0) {
            ann <- NULL
        } else {
            ann <- as.data.frame(lapply(ann, function(x) {
                if (nlevels(as.factor(x)) > 9) 
                  x else as.factor(x)
            }))
            # convert outlier scores back to numeric
            for (i in grep("_log2_outlier_score", colnames(ann))) {
                if (class(ann[, i]) == "factor") {
                  ann[, i] <- as.numeric(levels(ann[, i]))[ann[, i]]
                }
            }
        }
    } else {
        if (length(unique(ann)) > 1) {
            ann <- as.data.frame(ann)
            colnames(ann) <- show_pdata
            if (!grepl("_log2_outlier_score", show_pdata)) {
                ann <- as.data.frame(lapply(ann, function(x) {
                  if (nlevels(as.factor(x)) > 9) 
                    return(x) else return(as.factor(x))
                }))
            }
        } else {
            message(paste0("Column '", show_pdata, "' was excluded from annotation since they contained only a single value."))
            ann <- NULL
        }
    }
    return(ann)
}

organise_marker_genes <- function(object, k, p_val, auroc) {
    dat <- rowData(object)[, c(paste0("sc3_", k, "_markers_clusts"), paste0("sc3_", k, 
        "_markers_auroc"), paste0("sc3_", k, "_markers_padj"), "feature_symbol")]
    dat <- dat[dat[, paste0("sc3_", k, "_markers_padj")] < p_val & !is.na(dat[, paste0("sc3_", 
        k, "_markers_padj")]), ]
    dat <- dat[dat[, paste0("sc3_", k, "_markers_auroc")] > auroc, ]
    
    d <- NULL
    
    for (i in sort(unique(dat[, paste0("sc3_", k, "_markers_clusts")]))) {
        tmp <- dat[dat[, paste0("sc3_", k, "_markers_clusts")] == i, ]
        tmp <- tmp[order(tmp[, paste0("sc3_", k, "_markers_auroc")], decreasing = TRUE), ]
        d <- rbind(d, tmp)
    }
    
    return(d)
}

markers_for_heatmap <- function(markers) {
    res <- NULL
    for (i in unique(markers[, 1])) {
        tmp <- markers[markers[, 1] == i, ]
        if (nrow(tmp) > 10) {
            res <- rbind(res, tmp[1:10, ])
        } else {
            res <- rbind(res, tmp)
        }
    }
    
    return(res)
}
```

```{r fig.height=10, fig.width=10}
#plot_markers_trial <- function(object, k, auroc = 0.85, p.val = 0.01, show_pdata = NULL) {
  if (is.null(metadata(sce)$sc3$consensus)) {
    warning(paste0("Please run sc3_consensus() first!"))
    return(sce)
  }
  hc <- metadata(sce)$sc3$consensus[[as.character(4)]]$hc
  dataset <- get_processed_dataset(sce)
  if (!is.null(metadata(sce)$sc3$svm_train_inds)) {
    dataset <- dataset[, metadata(sce)$sc3$svm_train_inds]
  }

  add_ann_col <- FALSE
  ann <- NULL
  if (!is.null(NULL)) {
    ann <- make_col_ann_for_heatmaps(sce, NULL)
    if (!is.null(ann)) {
      add_ann_col <- TRUE
      # make same names for the annotation table
      rownames(ann) <- colnames(dataset)
    }
  }

  # get all marker genes
  markers <- organise_marker_genes(sce, 4, 0.01, 0.85)
  # get top 10 marker genes of each cluster
  markers <- markers_for_heatmap(markers)

  row.ann <- data.frame(clust_num = factor(markers[, 1], levels = unique(markers[, 1])))
  clust_names <- data.frame(clust_num = 1:4, Cluster = c("FAP/MAB", "Satellite cells", "Schwann cells", "Smooth muscle cells"))
  row.ann <- merge(row.ann, clust_names, by = "clust_num")
  row.ann <- row.ann["Cluster"]
  #row.ann <- data.frame(Cluster = sce$sc3_4_clusters)
  rownames(row.ann) <- markers$feature_symbol

  mat_colors <- list(Cluster = brewer.pal(4, "Set2"), genotype = c("#e41a1c", "#377eb8"))
  names(mat_colors$Cluster) <- unique(row.ann$Cluster)
  names(mat_colors$genotype) <- unique(ann$genotype)
  #names(Cluster) <- c("FAP/MAB", "Satellite cells", "Schwann cells", "Smooth muscle cells")
  #anno_colors <- list(Cluster = Cluster)

  do.call(pheatmap::pheatmap, c(list(dataset[markers$feature_symbol, , drop = FALSE],
                                     color = colorRampPalette(rev(brewer.pal(n = 11, name = "RdYlBu")))(100),
                                     show_colnames = FALSE,
                                     cluster_rows = FALSE, 
                                     cluster_cols = hc, 
                                     #cutree_cols = 4, 
                                     annotation_row = row.ann, 
                                     annotation_names_row = FALSE, 
                                     #gaps_row = which(diff(markers[, 1]) != 0), 
                                     cellheight = 14,
                                     cellwidth = 0.5,
                                     treeheight_col = 0),
                                     list(annotation_col = ann)[add_ann_col], 
                                     list(annotation_colors = mat_colors, 
                                     annotation_legend = FALSE,
                                     fontsize = 12,
                                     filename = "plots/heatmap/k4.pdf",
                                     width = 5,
                                     height = 8
                                     )))
#}
```

## Save RDS
```{r}
saveRDS(sce, file = "data/reads_qc_scran_SC3.rds")
```

# Session info
```{r}
sessionInfo()
```

